File: leetcode-js/047-permutations-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    nums.sort((a, b) => a - b);
    const res = [];
    helper(nums, 0, res);
    return res;
};

function helper(nums, index, res) {
    if (index === nums.length) {
        res.push(nums.slice());
        return;
    }

    for (let i = index; i < nums.length; i++) {
        if (i === index || nums[index] !== nums[i]) {
            swap(nums, index, i);
            helper(nums.slice(), index + 1, res);
        }
    }
}

function swap(nums, a, b) {
    const t = nums[a];
    nums[a] = nums[b];
    nums[b] = t;
}

console.log(permuteUnique([1,1,2,2]).length);

File: leetcode-java/047-permutations-ii.java
public class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        helper(nums, res, 0);
        return res;
    }

    void helper(int[] nums, List<List<Integer>> res, int index) {
        if (index == nums.length) {
            List<Integer> ans = new ArrayList<>();
            for (int x : nums) {
                ans.add(x);
            }
            res.add(ans);
            return;
        }

        for (int i = index; i < nums.length; i++) {
            if (i == index || nums[i] != nums[index]) {
                swap(nums, index, i);
                helper(nums.clone(), res, index + 1);
            }
        }
    }

    void swap(int[] nums, int a, int b) {
        int t = nums[a];
        nums[a] = nums[b];
        nums[b] = t;
    }
}
